### PR TITLE
Ensure modreq must exist on PE symbols that C# compiler expects them to have

### LINKED ISSUE #19792: Should referencing [IsReadOnly] signatures without an [InAttribute] modreq be allowed?
If a reference exists that has a virtual method with a ref-readonly parameter and [IsReadOnly] attribute, but no [IsConst] required modifier: Per feature design this is not possible, but it can be created by external tools/compilers. Should it be imported and used in the compilation? or should we mark it as a bad symbol and generate a use site error for it?

This issue should track design decisions taken around this issue, and how to fix it.
Issue was raised in #19658, where the PR for now allows it and added tests to verify the behavior.

cc @AlekseyTs @VSadov

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #19792
cc @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet-bot test this

**Comment by jaredpar** [comment]
Shouldn't we target dev15.7.x with this change?

**Comment by mavasani** [comment]
@dotnet-bot retest this please

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler this is now ready

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler this is ready for review

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 1]
```diff
@@ -143,10 +143,30 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
             PEModuleSymbol moduleSymbol,
             PEMethodSymbol containingSymbol,
             int ordinal,
-            ParamInfo<TypeSymbol> parameter,
+            ParamInfo<TypeSymbol> parameterInfo,
+            bool isreturnParam,
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: `isReturn...`

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 2]
```diff
@@ -158,17 +178,31 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
...
             PEPropertySymbol containingSymbol,
             int ordinal,
             ParameterHandle handle,
-            ParamInfo<TypeSymbol> parameter,
+            ParamInfo<TypeSymbol> parameterInfo,
             out bool isBad)
         {
-            return Create(moduleSymbol, containingSymbol, ordinal, parameter.IsByRef, parameter.RefCustomModifiers, parameter.Type, handle, parameter.CustomModifiers, out isBad);
+            var parameter = Create(moduleSymbol, containingSymbol, ordinal, parameterInfo.IsByRef, parameterInfo.RefCustomModifiers, parameterInfo.Type, handle, parameterInfo.CustomModifiers, out isBad);
+            var inAttributeModreq = parameter.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: There are several instances of this expression. Consider extracting `HasInAttribute(ImmutableArray<ICustomModifier>)` helper method, or extract a static delegate.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Would it make sense to move the modreq check to the caller rather than passing in an `isReturnParam` argument? If so, then consider moving the modreq check for the `PEPropertySymbol` parameter case as well.

File: docs/compilers/CSharp/Compiler Breaking Changes - post VS2017.md [hunk 3]
```diff
@@ -95,3 +95,5 @@ Example: `Func<int> f = default(TypedReference).GetHashCode; // new error CS0123
     }
     ```
     This is changed in 15.6 to now produce an error that the variable is not definitely assigned.
+
+- VS 15.7 : https://github.com/dotnet/roslyn/issues/19792 C# compiler will now reject [IsReadOnly] symbols that should have an [InAttribute] modreq, but don't.
```
Comment (jcouv) [inline docs/compilers/CSharp/Compiler Breaking Changes - post VS2017.md]: Nit: please spell out "Visual Studio 2017 version 15.5" (branding guidelines)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 4]
```diff
@@ -143,10 +143,30 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
             PEModuleSymbol moduleSymbol,
             PEMethodSymbol containingSymbol,
             int ordinal,
-            ParamInfo<TypeSymbol> parameter,
+            ParamInfo<TypeSymbol> parameterInfo,
+            bool isreturnParam,
             out bool isBad)
         {
-            return Create(moduleSymbol, containingSymbol, ordinal, parameter.IsByRef, parameter.RefCustomModifiers, parameter.Type, parameter.Handle, parameter.CustomModifiers, out isBad);
+            var parameter = Create(moduleSymbol, containingSymbol, ordinal, parameterInfo.IsByRef, parameterInfo.RefCustomModifiers, parameterInfo.Type, parameterInfo.Handle, parameterInfo.CustomModifiers, out isBad);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Nit: consider breaking long line.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 5]
```diff
@@ -143,10 +143,30 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
...
+            var inAttributeModreq = parameter.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >var [](start = 12, length = 3)

Please spell out the type.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 6]
```diff
@@ -143,10 +143,30 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
...
+
+            if (isreturnParam)
+            {
+                // A return parameter of type In should always have this modreq, and vice versa.
+                isBad |= (parameter.RefKind == RefKind.RefReadOnly) != inAttributeModreq;
+            }
+            else if (parameter.RefKind == RefKind.In)
+            {
+                // An in parameter should not have this modreq, unless the method was virtual or abstract.
+                isBad |= (containingSymbol.IsVirtual || containingSymbol.IsAbstract) != inAttributeModreq;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

Should we use IsMetadataVirtual here instead?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >Create [](start = 28, length = 6)

It looks like we are duplicating a lot of code, can we push validation into the common Create helper?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 7]
```diff
@@ -158,17 +178,31 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
...
             out bool isBad)
         {
-            return Create(moduleSymbol, containingSymbol, ordinal, parameter.IsByRef, parameter.RefCustomModifiers, parameter.Type, handle, parameter.CustomModifiers, out isBad);
+            var parameter = Create(moduleSymbol, containingSymbol, ordinal, parameterInfo.IsByRef, parameterInfo.RefCustomModifiers, parameterInfo.Type, handle, parameterInfo.CustomModifiers, out isBad);
+            var inAttributeModreq = parameter.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
+
+            if (parameter.RefKind == RefKind.In)
+            {
+                // An in parameter should not have this modreq, unless the method was virtual or abstract.
+                isBad |= (containingSymbol.IsVirtual || containingSymbol.IsAbstract) != inAttributeModreq;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

It looks like we should also check IsOverride

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

It looks like this method is called form constructor of PEPropertySymbol, the symbol might not be fully initialized. I think we should avoid calling any APIs through containingSymbol. We can calculate this information at the call site in a safe manner and pass it in.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 5 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: `InAttribute` is the type name. Did you mean leaving a comment?

---
In reply to: [166435944](https://github.com/dotnet/roslyn/pull/24524#discussion_r166435944) [](ancestors = 166435944)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 6 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: it is not available for properties.

---
In reply to: [166437575](https://github.com/dotnet/roslyn/pull/24524#discussion_r166437575) [](ancestors = 166437575)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 2 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: refactored and removed duplicate

---
In reply to: [165710734](https://github.com/dotnet/roslyn/pull/24524#discussion_r165710734) [](ancestors = 165710734)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 7 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: these getters should guard against that. correct?

---
In reply to: [166439938](https://github.com/dotnet/roslyn/pull/24524#discussion_r166439938) [](ancestors = 166439938)

Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: AFAIU, this is checked when we load the override that it matches the overridden correctly, so we don't need that extra check here.

---
In reply to: [166439052](https://github.com/dotnet/roslyn/pull/24524#discussion_r166439052) [](ancestors = 166439052)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 1 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: I'd rather move it to the inner helper, rather than enumerating the modifiers twice.

---
In reply to: [165716449](https://github.com/dotnet/roslyn/pull/24524#discussion_r165716449) [](ancestors = 165716449)

**Comment by OmarTawfik** [comment]
@AlekseyTs @jcouv @cston feedback addressed.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs [hunk 8]
```diff
@@ -79,9 +78,12 @@ private enum Flags : byte
             }
             else
             {
-                result = new PEPropertySymbolWithCustomModifiers(moduleSymbol, containingType, handle, getMethod, setMethod, propertyParams, metadataDecoder, out isBad);
+                result = new PEPropertySymbolWithCustomModifiers(moduleSymbol, containingType, handle, getMethod, setMethod, propertyParams, metadataDecoder);
             }
 
+            // A property should always have this modreq, and vice versa.
+            var isBad = (result.RefKind == RefKind.In) != result.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
+
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs]: Please extract a `HasInAttribute(ImmutableArray<ICustomModifier>)` helper.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs [hunk 9]
```diff
@@ -250,6 +250,769 @@ public void TestFieldCustomModifierCount()
             Assert.Equal(0, CustomModifierCount(@class.GetMember<FieldSymbol>("field00")));
         }
 
+        [Fact]
+        public void RejectIsReadOnlySymbolsThatShouldHaveInAttributeModreqButDoNot_Delegates_Parameters()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: Consider shortening the test names. Perhaps `MissingInAttributeModreq_Delegates_Parameters`

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 10]
```diff
@@ -272,14 +279,39 @@ private bool HasNameInMetadata
...
             {
-                return new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+                parameter = new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+            }
+            else
+            {
+                parameter = new PEParameterSymbolWithCustomModifiers(moduleSymbol, containingSymbol, ordinal, isByRef, refCustomModifiers, type, handle, customModifiers, out isBad);
+            }
+
+            var inAttributeModreq = parameter.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >var [](start = 12, length = 3)

Please spell out the type instead of using ```var```.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 11]
```diff
@@ -272,14 +279,39 @@ private bool HasNameInMetadata
...
+
+            if (isReturn)
+            {
+                // A RefReadOnly return parameter should always have this modreq, and vice versa.
+                isBad |= (parameter.RefKind == RefKind.RefReadOnly) != inAttributeModreq;
+            }
+            else if (parameter.RefKind == RefKind.In)
+            {
+                // An in parameter should not have this modreq, unless the containing symbol was virtual or abstract.
+                isBad |= (containingSymbol.IsVirtual || containingSymbol.IsAbstract) != inAttributeModreq;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

I am not comfortable with usage of these APIs for properties based on the same reasons that I outlined during the previous review pass.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: I would suggest to get IsMetadataVirtual from the getter and pass that in instead of trying to calculate it here.

---
In reply to: [167048153](https://github.com/dotnet/roslyn/pull/24524#discussion_r167048153) [](ancestors = 167048153)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

As I mentioned during the previous review pass, it looks like this will not properly handle an method or a property with IsOverride == true. Please make sure to add a test for this scenario.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Symbols/CustomModifier.cs [hunk 12]
```diff
@@ -122,4 +123,12 @@ public override bool Equals(object obj)
             }
         }
     }
+
+    internal static class CustomModifierExtensions
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/Symbols/CustomModifier.cs]: note to self: perhaps move to `CustomModifierUtils`?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 13]
```diff
@@ -266,20 +275,46 @@ private bool HasNameInMetadata
...
+
             if (customModifiers.IsDefaultOrEmpty && refCustomModifiers.IsDefaultOrEmpty)
             {
-                return new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+                parameter = new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+            }
+            else
+            {
+                parameter = new PEParameterSymbolWithCustomModifiers(moduleSymbol, containingSymbol, ordinal, isByRef, refCustomModifiers, type, handle, customModifiers, out isBad);
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Consider using ternary operator: `var parameter = (...) ? ...: ...;`

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs [hunk 14]
```diff
@@ -133,7 +135,14 @@ private enum Flags : byte
...
             bool isBad;
-            _parameters = GetParameters(moduleSymbol, this, propertyParams, setMethodParams ?? getMethodParams, out isBad);
+            if (setMethodParams is null)
+            {
+                _parameters = GetParameters(moduleSymbol, this, propertyParams, getMethodParams, getMethod.IsMetadataVirtual(), out isBad);
+            }
+            else
+            {
+                _parameters = GetParameters(moduleSymbol, this, propertyParams, setMethodParams, setMethod.IsMetadataVirtual(), out isBad);
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEPropertySymbol.cs]: Consider using ternary operator.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs [hunk 15]
```diff
@@ -250,6 +250,1915 @@ public void TestFieldCustomModifierCount()
...
+{
+    public int M(RefTest obj) => obj.M(0);
+}", references: new[] { reference }).VerifyDiagnostics(
+                // (4,38): error CS0570: 'RefTest.M(in int)' is not supported by the language
+                //     public int M(RefTest obj) => obj.M(0);
+                Diagnostic(ErrorCode.ERR_BindToBogus, "M").WithArguments("RefTest.M(in int)").WithLocation(4, 38));
+        }
+
+        [Fact]
+        public void MissingInAttributeModreq_Method_Parameters_Virtual_ModOpt()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: Would it be sufficient to have one `_ModOpt` version of the tests for methods and one for properties. 

Also, some of the suffixes are `_Modopt` and some are `_ModOpt`.

**Comment by OmarTawfik** [comment]
>     public class LoadCustomModifiers : CSharpTestBase

note to self: Add tests where modreq exists on override, but parent has none or has a modopt

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs:12 in 3f4518b. [](commit_id = 3f4518b4ef00c80928ab333b61df67b153dd6ae7, deletion_comment = False)

**Comment by OmarTawfik** [comment]
>     public class LoadCustomModifiers : CSharpTestBase

note to self: for overriding methods, make sure it has a use site error through the symbol API.

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs:12 in 3f4518b. [](commit_id = 3f4518b4ef00c80928ab333b61df67b153dd6ae7, deletion_comment = False)

**Comment by OmarTawfik** [comment]
>     public class LoadCustomModifiers : CSharpTestBase

also, verify IL.

---
In reply to: [364577700](https://github.com/dotnet/roslyn/pull/24524#issuecomment-364577700) [](ancestors = 364577700)

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs:12 in 3f4518b. [](commit_id = 3f4518b4ef00c80928ab333b61df67b153dd6ae7, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs (hunk 15 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: @cston can you please clarify?
fixed the naming issue.

---
In reply to: [167270920](https://github.com/dotnet/roslyn/pull/24524#discussion_r167270920) [](ancestors = 167270920)

Comment (cston) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: There are twenty pairs of tests where each pair has one test with missing attribute and one test with `modopt` rather than `modreq`. That is a large number of tests. But `HasInAttributeModifier` checks for both the attribute and `modreq` so it seems like overkill to have two tests for each. It feels like one or two tests for `modopt` should be sufficient.

Comment (cston) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: Is there a test where both `modopt` and `modreq` are included?

Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadCustomModifiers.cs]: There are tests where a virtual method in a parent class has a modreq, but the overriding method in a child class has a modopt. I'm adding the opposite right now.

---
In reply to: [167356516](https://github.com/dotnet/roslyn/pull/24524#discussion_r167356516) [](ancestors = 167356516)

**Comment by OmarTawfik** [comment]
@AlekseyTs comments addressed.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 16]
```diff
@@ -266,20 +275,39 @@ private bool HasNameInMetadata
...
+
+            if (isReturn)
+            {
+                // A RefReadOnly return parameter should always have this modreq, and vice versa.
+                isBad |= (parameter.RefKind == RefKind.RefReadOnly) != hasInAttributeModifier;
+            }
+            else if (parameter.RefKind == RefKind.In)
+            {
+                // An in parameter should not have this modreq, unless the containing symbol was virtual or abstract.
+                isBad |= isContainingSymbolVirtual != hasInAttributeModifier;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: >isContainingSymbolVirtual != hasInAttributeModifier [](start = 25, length = 51)

Do we want to be that restrictive here? It feels like we do know what it means to have the modifier even if the method is not virtual. So if we might say that the method is bad only if the method is virtual and the modifier is missing, but it is fine to have a modifier for non-virtual method.

**Comment by AlekseyTs** [comment]
>     public class LoadInAttributeModifier : CSharpTestBase

It looks like there is no single reference to IsOverride property in this file. How do I know that a property and a method with IsOverride == true are covered?

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadInAttributeModifier.cs:11 in 180fbc8. [](commit_id = 180fbc8cb31c36885b1a7e7e3c77cbe78082db72, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         [ConditionalFact(typeof(ClrOnly))] // https://github.com/mono/mono/issues/6936

Instead of restricting this to CLROnly could we disabled it specifically for Mono instead?

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadInAttributeModifier.cs:316 in 180fbc8. [](commit_id = 180fbc8cb31c36885b1a7e7e3c77cbe78082db72, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 16 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: We use the modifier to indicate a specific combination of metadata. Invalidating this specific combination means that it is something out of our design, and that some IL user is misusing the modreq. I believe such signatures should be disallowed.

---
In reply to: [167711399](https://github.com/dotnet/roslyn/pull/24524#discussion_r167711399) [](ancestors = 167711399)

**Comment by OmarTawfik** [comment]
>         [ConditionalFact(typeof(ClrOnly))] // https://github.com/mono/mono/issues/6936

Unless I'm missing something, this is the flag we use to not run on mono. Please see:
src\Test\Utilities\Portable\Assert\ConditionalFactAttribute.cs
```csharp
    public class ClrOnly : ExecutionCondition
    {
        public override bool ShouldSkip => MonoHelpers.IsRunningOnMono();
        public override string SkipReason => "Test not supported on Mono";
    }
```

---
In reply to: [365091162](https://github.com/dotnet/roslyn/pull/24524#issuecomment-365091162) [](ancestors = 365091162)

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadInAttributeModifier.cs:316 in 180fbc8. [](commit_id = 180fbc8cb31c36885b1a7e7e3c77cbe78082db72, deletion_comment = False)

**Comment by OmarTawfik** [comment]
>     public class LoadInAttributeModifier : CSharpTestBase

Added the required asserts

---
In reply to: [365090757](https://github.com/dotnet/roslyn/pull/24524#issuecomment-365090757) [](ancestors = 365090757)

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadInAttributeModifier.cs:11 in 180fbc8. [](commit_id = 180fbc8cb31c36885b1a7e7e3c77cbe78082db72, deletion_comment = False)

**Comment by OmarTawfik** [comment]
@AlekseyTs Addressed latest feedback.

**Comment by OmarTawfik** [comment]
@jaredpar can we get sign off to merge?

**Comment by jaredpar** [comment]
approved
